### DISCUSSION TITLE
Scene without action triggers on other scenes.

### DISCUSSION THREAD

**Discussion by JustMe-NL** [OMOTE-Community/OMOTE-Firmware#76] [Ideas]
What do I mean with a scene without actiontrigger?

It is a scene you can open & close without triggering the respectively close & open commands in the previous scene. Let me explain:

Suppose we are in the "watch a DVD" scene, so the TV, the DVD player and the Surround system are all on. Now you want to issue a command to your smarthome to switch the lights to fit the mood. It would be nice if you could press a button to switch to a smarthome scene, scroll to the lightscenes gui and select the mood you want, then press the button again to return to your pervious scene. 

But this would mean that with the current setup, that leaving the "watch a DVD scene" would issue a turn of DVD command and switch TV back to Cable before entering the smarthome scene and switching them back on after leaving the smarthome scene and returning to the "watch a DVD" scene.

With the Harmony One you can switch to a device from within a scene and back to the scene without triggering the scene setup-commands.

It is a bit of a challenge because it would mean that the scenehandler should check the previous or next scene to see if it needs to trigger commands and it should be able to "toggle" the "special" scene with a executeCommand.

**Comment by KlausMu** [comment]
If I understand you correctly, then I think this could be achieved in two ways:

- put the smarthome gui(s) in the scene specific gui list (could be put in all scene specific gui lists)
- or, more clever, put it/them in the main_gui_list

Because you can always navigate back to the main_gui_list (with the scene selection gui as first gui) and again navigate to the scene specific list **without** triggering the start sequence again. If you select the same scene again, the start sequence is not repeated. Only if you do a long press on the scene button which is already active, a so called "force" command is triggered (this is explained in the wiki).

Doesn't this fit your needs? If not, could you please explain in more detail why it does not fit?

So If you compare it to Harmony, I would put the device guis in the main gui list.

**Reply by JustMe-NL** [reply]
Yes, that could work. But what is missing is the ability to return to the previous scene by using a single hardware button to call the scene_selector and exit it again, sort of like a SCENE_PREVIOUS.

And alas not a complete replacement for the "Harmony devices" function as well, unless I create a gui per device (doable) and set the hardware keys to correspond the shown gui. But currently hardware keys are linked to scenes and not to gui's right? Could I utilize the create/destroy gui calls to backup and restore the hardwarekey mappings? (Allthough I do guess I need a "gui_memoryOptimizer_isGUInameActive" next to the "gui_memoryOptimizer_isGUInameInMemory" then).

**Reply by KlausMu** [reply]
So would you say you liked to have more support of hardware keys?
Because all of your feature requests could already be fullfilled by the gui, right?

- navigate to a device specific gui, keeping the context of the currently active scene
- send commands to that device from the gui
- go back to the scene specific gui list (via selecting the specific scene again, currently not by "SCENE_NAVIGATE_TO_GUI_LIST")

BTW, I wouldn't call it SCENE_PREVIOUS, because it is not the previous scene, but the still active scene.

You would like to have

1. key bindings based on the currently active gui (not only based on the scene)
2. a command like "SCENE_NAVIGATE_TO_GUI_LIST" to activate the scene specific gui list (when being in the main_gui_list), this command could be bind to a hardware key

For 1. the sceneRegistry had to be extended (get_command_short() etc.)
2. shouldn't be too complicated.

So is this what you would like to have?

To be honest, I never used the device pages on Harmony. I never needed them.

Could you please explain in more detail which kind of guis you would like to have in that way? And what the hardware keys should do? Are these guis really for the devices involved in that scene, or are they for something completely different?

**Reply by JustMe-NL** [reply]
Sounds about right, let me try with a few userstories:

- As an user of the remote I would like to push a hardwarebutton to get the screen(s) to control my Smarthome from every scene and return to said scene by pushing the same hardwarebutton so I can switch back and forth without any thought about what I was doing. I would like it so that the push of the button takes me directly to the gui screen that controls the smarthome.

- As an user of the remote I would like to be able to switch to a specific device to issue a command, either through a gui or with hardwarebuttons and easily return to the active scene after I'm done so I can fix any IR command the scene switch might have missed of to correct some obscure setting in one of the devices the current scene does not provide a command for. If I need to swipe some screens to get to the device in question, that is not really a problem although it would be nice to get there by selecting  the device through a (a few) button pushes on the gui.

Given these stories, (1) key bindings based on the currently active gui would help a lot and (2) a command to navigate to a specific gui within the list would help.

**Reply by KlausMu** [reply]
What I would like to have is a general purpose concept. So let's try it.
Currenty we have the main gui list and scene specific gui lists.

I understand the need for device specific guis to do some low level stuff on the devices:

- one or more gui per device, not scene specific, but with gui specific key bindings
- a good place where to put these device guis
- a simple way to get to the device guis and back to the scene guis

Something similiar could be done with the Smart Home guis. I would put them into the main gui list.

![20240415_112514_cr](https://github.com/CoretechR/OMOTE/assets/14290221/273c733d-5081-401e-86b4-ea602b2cb7e9)

So overall the new features could be

- let device guis be available left to the scene selection gui
- allow gui specific key bindings (in red, would overwrite scene specific key bindings)
- let GUIs define commands so that you can directly jump to that GUI (in green, ~~only possible if they are in the main gui list~~)
- introduce the command "SCENE_BACK_TO_PREVIOUS_GUI_LIST": jumps to the last active GUI from the other list ~~introduce the command "SCENE_BACK_TO_SCENE_GUI_LIST": jumps to the last active GUI from the scene list (in the example one of A1, A2, or A3). Only available if you are not in the scene gui list~~

How does this sound?

**Reply by JustMe-NL** [reply]
That sounds like it ticks all of the boxes!

**Comment by KlausMu** [comment]
The changes are now available in branch "improved_navigation" and as [PR](https://github.com/CoretechR/OMOTE/pull/81).

Please see the updated [Wiki ](https://github.com/CoretechR/OMOTE/wiki/02-How-to-understand-and-modify-the-software) and do a search in the Wiki for
- GUI_SMARTHOME_ACTIVATE 
- SCENE_BACK_TO_PREVIOUS_GUI_LIST
- "The same is possible for GUIs. They can also override key definitions ..."

Devices can simply be put left to the scene selection gui with `main_gui_list = {...}` in `main.cpp`.
Please see the example with the new gui for the yamahaAmp.

Could you please test it and provide feedback if you find any issue?`

**Reply by JustMe-NL** [reply]
Sorry it took a while to respond but I was chasing a non-bug around. I copied all the changes over to my code but the keys did not work. The gotcha is that you really need to register the guis before the scenes! After I found the "bug", it's all so obvious since all the commands are indexed at runtime.

Seems to work great after I changed the order in main (may need a warning in the PR for this one).

I'll keep on testing and playing around with it and post my findings, haven't tested adding the devices to the scene selection yet.

**Reply by KlausMu** [reply]
Ok, sorry for the confusion.

I think I will add the same code to the gui registry as already being in the scene registry, so that it doesn't matter if you first register the guis or the scenes.

If it does not work, it is really hard to find.

**Reply by KlausMu** [reply]
Ok, I added the code to the branch.
